### PR TITLE
feat(auth): confirm user attribute

### PR DESCRIPTION
**Pull request by banji180**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*
Added implementation for confirm user attribute

*How did you test these changes?*
(Please add a line here how the changes were tested)

- [x] Added Unit Tests
- [x] Tested on frontend (Add Screenshots)
- [x] Successful logs (Attach them to the PR)

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/main/java/com/amplifyframework/auth/AuthException.java [hunk 1]
```diff
@@ -635,6 +635,96 @@ public UserCancelledException(String message, Throwable cause, String recoverySu
...
+        }
+    }
+
+    /**
+     * Could not perform the operation due to invalid sms role trust relationship.
+     */
+    public static class InvalidSmsRoleTrustRelationshipException extends AuthException {
+        private static final long serialVersionUID = 1L;
+        private static final String MESSAGE = "Invalid SMS Role Trust Relationship error.";
+        private static final String RECOVERY_SUGGESTION =
```
Comment (gpanshu) [inline core/src/main/java/com/amplifyframework/auth/AuthException.java:680]: I believe you have the same class in PR 1884 ?

Comment (banji180) [reply core/src/main/java/com/amplifyframework/auth/AuthException.java:680]: this branch is based off the updateuserattribute branch. The updateuserattribute branch needs to be merged first

**Review by gpanshu — DISMISSED** [review]

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 2]
```diff
@@ -785,15 +793,48 @@ internal class RealAWSCognitoAuthPlugin(
...
     ) {
-        TODO("Not yet implemented")
+        authStateMachine.getCurrentState { authState ->
+            when (authState.authNState) {
+                // Check if user signed in
+                is AuthenticationState.SignedIn -> {
+                    GlobalScope.launch {
+                        try {
+                            val attributes = listOf(attribute)
+                            val results = updateUserAttributes(attributes?.toMutableList(), null)
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:828]: `attribute` is non nullable, so `listOf(attribute)` will never be null.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 3]
```diff
@@ -785,15 +793,48 @@ internal class RealAWSCognitoAuthPlugin(
...
-        TODO("Not yet implemented")
+        authStateMachine.getCurrentState { authState ->
+            when (authState.authNState) {
+                // Check if user signed in
+                is AuthenticationState.SignedIn -> {
+                    GlobalScope.launch {
+                        try {
+                            val attributes = listOf(attribute)
+                            val userAttributesOptions = options as? AWSCognitoAuthUpdateUserAttributesOptions
+                            val results = updateUserAttributes(attributes?.toMutableList(), userAttributesOptions)
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:804]: `attribute` is non nullable, so `listOf(attribute)` will never be null.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 4]
```diff
@@ -802,15 +843,115 @@ internal class RealAWSCognitoAuthPlugin(
...
+                        }
+                    }
+                }
+                else -> onError.accept(AuthException.InvalidStateException())
+            }
+        }
+    }
+
+    private suspend fun updateUserAttributes(
+        attributes: MutableList<AuthUserAttribute>,
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:887]: This does not have to be a mutableList.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 5]
```diff
@@ -785,15 +793,48 @@ internal class RealAWSCognitoAuthPlugin(
         onSuccess: Consumer<AuthUpdateAttributeResult>,
         onError: Consumer<AuthException>
     ) {
-        TODO("Not yet implemented")
+        authStateMachine.getCurrentState { authState ->
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:796]: Checks for SM's currentState can be moved to private `updateUserAttributes` for all the API variants.

**Comment by div5yesh** [comment]
This PR contains changes from #1884. Either this should merge into #i884 or it should only contain changes related to `confirmUserAttribute`.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 6]
```diff
@@ -836,7 +977,31 @@ internal class RealAWSCognitoAuthPlugin(
...
+                is AuthenticationState.SignedIn -> {
+                    GlobalScope.launch {
+                        try {
+                            val accessToken = getSession().userPoolTokens.value?.accessToken
+                            val verifyUserAttributeRequest = VerifyUserAttributeRequest.invoke {
+                                this.accessToken = accessToken
+                                this.attributeName = attributeKey.keyString
+                                this.code = confirmationCode
+                            }
+                            val verifyUserAttributeResponse = authEnvironment.cognitoAuthService
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:992]:
```suggestion
                            authEnvironment.cognitoAuthService
```

File: aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt [hunk 7]
```diff
@@ -472,4 +478,177 @@ class RealAWSCognitoAuthPluginTest {
...
+        }
+
+        plugin.confirmUserAttribute(
+            AuthUserAttributeKey.email(),
+            "000000",
+            onSuccess,
+            onError
+        )
+
+        assertTrue { listenLatch.await(60, TimeUnit.SECONDS) }
```
Comment (div5yesh) [inline aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt:650]: Also verify if Amplify API response matches expected response.

File: core/src/main/java/com/amplifyframework/auth/AuthException.java (hunk 1 above)
Comment (div5yesh) [reply core/src/main/java/com/amplifyframework/auth/AuthException.java:680]: You can change the base to 1884's branch and when 1884 is merged GH will automatically update this PR's base as dev-preview.

**Review by div5yesh — DISMISSED** [review]
Few minor things. Most of the comments are relevant for #1884.

File: aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt (hunk 7 above)
Comment (banji180) [reply aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt:1119]: The Amplify API response is an empty VerifyUserAttributeResponse

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 8]
```diff
@@ -1024,7 +1025,31 @@ internal class RealAWSCognitoAuthPlugin(
...
+                            val verifyUserAttributeRequest = VerifyUserAttributeRequest.invoke {
+                                this.accessToken = accessToken
+                                this.attributeName = attributeKey.keyString
+                                this.code = confirmationCode
+                            }
+                            authEnvironment.cognitoAuthService
+                                .cognitoIdentityProviderClient?.verifyUserAttribute(
+                                    verifyUserAttributeRequest
+                                )
+                            onSuccess.call()
```
Comment (sktimalsina) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:1044]: Isn't `verifyUserAttributeRequest` a suspended function? 
I am afraid `onSuccess.call()` can run concurrently with `verifyUser..` and return false positive. You should chain the `onSuccess` with `verifyUser..`.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 9]
```diff
@@ -1024,7 +1025,31 @@ internal class RealAWSCognitoAuthPlugin(
...
         onError: Consumer<AuthException>
     ) {
-        TODO("Not yet implemented")
+        authStateMachine.getCurrentState { authState ->
+            when (authState.authNState) {
+                // Check if user signed in
+                is AuthenticationState.SignedIn -> {
+                    GlobalScope.launch {
+                        try {
+                            val accessToken = getSession().userPoolTokens.value?.accessToken
```
Comment (sktimalsina) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:1034]: what happens when accessToken is null? looks like we are missing test for it.

File: aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt [hunk 10]
```diff
@@ -1052,4 +1054,70 @@ class RealAWSCognitoAuthPluginTest {
         )
         coVerify(exactly = 0) { onError.accept(any()) }
     }
+
+    @Test
+    fun `confirm user attribute fails when not in SignedIn state`() {
```
Comment (sktimalsina) [inline aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt:1059]: Can we add a test for when cognito throws an error?

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt (hunk 9 above)
Comment (banji180) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:1040]: fixed

File: aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt (hunk 10 above)
Comment (banji180) [reply aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPluginTest.kt:1059]: yes done

**Review by sktimalsina — APPROVED** [review]

**Review by gpanshu — APPROVED** [review]

**Pull request merged**